samcli/lib/telemetry/metric.py

```diff
@@ -263,6 +279,89 @@ def _get_project_details(hook_name: str, template_dict: Dict) -> ProjectDetails:
     )
 
 
+def _get_stack_trace_info(exception: Exception) -> Tuple[Optional[str], Optional[str]]:
+    """
+    Takes an Exception instance and extracts the following:
+      1. Stack trace in a readable string format with user-sensitive paths cleaned
+      2. Exception mesage including the fully-qualified exception name and value
+
+    Parameters
+    ----------
+    exception : Exception
+        Exception instance
+
+    Returns
+    -------
+    (str, str)
+        (stack trace, exception message)
+    """
+    stack_trace, exception_msg = None, None
+    try:
+        tb_exception = traceback.TracebackException.from_exception(exception)
+        _clean_traceback_recursively(tb_exception)
+        stack_trace = "".join(list(tb_exception.format(chain=True)))
+        exception_msg = list(tb_exception.format_exception_only())[-1]
+    except Exception as ex:
+        LOG.debug("failed to retrieve stack trace or exception message: %s", ex)
+
+    return (stack_trace, exception_msg)
+
+
+def _clean_traceback_recursively(tbexp: traceback.TracebackException) -> None:
+    """
+    Cleans stack trace recursively by going through both cause and context
+    We don't handle RecursionError because traceback.TracebackException prevents it
+
+    Parameters
+    ----------
+    tbexp : traceback.TracebackException
+        TracebackException instance
+    """
+    # __cause__ is populated by using `raise ... from ...` - PEP 3134
+    if tbexp.__cause__:
+        _clean_traceback_recursively(tbexp.__cause__)
+
+    # __context__ is populated by raising an exception in an except block - PEP 3134
+    if tbexp.__context__:
+        _clean_traceback_recursively(tbexp.__context__)
+
+    _clean_stack_summary_paths(tbexp.stack)
+    # redacting entire exception message
+    tbexp._str = "<REDACTED>"  # type: ignore
+
+
+def _clean_stack_summary_paths(stack_summary: traceback.StackSummary) -> None:
+    """
+    Cleans the user-sensitive paths contained within a StackSummary instance
+
+    Parameters
+    ----------
+    stack_summary : traceback.StackSummary
+        StackSummary instance
+    """
+    for frame in stack_summary:
+        path = frame.filename
+        separator = "\\" if "\\" in path else "/"
+
+        # Case 1: If "site-packages" is found within path, replace its leading segment with: /../ or \..\
```

Comment:
Not blocking with respect to logic, but wanted to add from readability and maintainability perspective.

Each of the logic of if section can be put into a class with singular apply method.

This way below looks cleaner.

```python
class FrameMapper:
      def apply():
           pass

frame_mappers = [SitePackageFrameMapper(), SamcliFrameMapper(), LastFileFrameMapper()]
```

Now below looks simpler and scales well if there are additional mappers that come into play without resulting in many if statements.
```python
for frame in stack_summary:
     for frame_mapper in frame_mappers:
          frame.filename = frame_mapper.apply(*args, **kwargs)
```